# DATA CONSITENCY CHECKS

# 1. Importing pandas, NumPy, and os libraries

# 2. Importing 'products.csv' and 'orders_wrangled.csv'

# 3. Creating a dataframe with mixed type column and replacing it with string

# 4. Finding missing values in 'df_prods'

# 5. Finding mean and median values in 'df_prods'

# 6. Removing missing values in 'df_prods'

# 7. Finding and dropping duplicates in 'df_prods_clean'

# 8. Exporting 'products_checked' without duplicates

# 9. Data Consistency Checks (Task)

## 9.1. Describing 'df_prods'

## 9.2. Checking for mixedtype data in 'df_ords'

## 9.3. Addressing missing values in 'df_ords'

## 9.4. Checking for duplicates in 'df_ords'

## 9.5. Exporting 'df_ords' as 'orders_checked'

In [40]:
# ----------------------------------------------

# 1. Importing pandas, NumPy, and os libraries

In [1]:
# Importing the pandas, NumPy, and os libraries:

import pandas as pd
import numpy as np
import os

# 2. Importing 'products.csv' and 'orders_wrangled.csv'

In [2]:
# Defining path:

path = r'/Users/elena/Documents/Data Analytics Course/4. Python fundamentals for Data Analysts/Instacart Basket Analysis'

In [3]:
# Importing products.csv:

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'products.csv'))

In [4]:
# Importing orderswrangled.csv:

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared data', 'orders_wrangled.csv'))

# 3. Creating a dataframe with mixed type column and replacing it with string

In [5]:
# Creating a dataframe:

df_test = pd.DataFrame()

In [7]:
# Creating a mixed type column:

df_test['mix'] = ['a', 'b', 1, True]
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Checking for mixed types in df_test:

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
# Replacing mixed type with string type in df_test:

df_test['mix'] = df_test['mix'].astype('str')

# 4. Finding missing values in 'df_prods'

In [11]:
# Finding missing values in df_prods:

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# Creating a subset with missing values only:

df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# 5. Finding mean and median values in 'df_prods'

In [13]:
# Finding the mean values for the columns with missing values:

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [14]:
# Finding the median value in df_prods:

df_prods.median()

/var/folders/_w/cktj2zv50h34n2spc6h0z5pc0000gp/T/ipykernel_76375/951194849.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_prods.median()


product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [15]:
# Finding the number of rows and columns in df_prods:

df_prods.shape

(49693, 5)

# 6. Removing missing values in 'df_prods'

In [16]:
# Creating a new df_prods_clean without missing values:

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [17]:
df_prods_clean.shape

(49677, 5)

# 7. Finding and dropping duplicates in 'df_prods_clean'

In [18]:
# Finding duplicates in df_prods_clean:

df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
# Dropping duplicates in df_prods_clean:

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

# 8. Exporting 'products_checked' without duplicates

In [21]:
# Exporting df_prods_clean_no_dups:

df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data','Prepared Data','products_checked.csv'))

# 9. Data Consistency Checks (Task)

## 9.1. Describing 'df_prods'

In [22]:
# Item 2. Running df.describe() on df_prods:

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [23]:
    ## Interpreting the output of "df_prods.describe()":
    
df_prods.median()

/var/folders/_w/cktj2zv50h34n2spc6h0z5pc0000gp/T/ipykernel_76375/3922971161.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_prods.median()


product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

#### 1. The max value for "product_id" (49688) is 5 points lower than the count (49693) meaning that there might be duplicates or missing values in the column
2. The max value for the "prices" being 99999.000000, while the median is only 7.1, gives us an idea that this 99999.00000 might be an outlier and needs to be dealt with accordingly.
3. The min value for "prices" is exactly 1.000000, which is suspicious, taking into account that there are usually multiple items in stores that are priced under $1.

## 9.2. Checking for mixedtype data in 'df_ords'

In [24]:
# Item 3. Checking for mixed-type data in 'df_ords':

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### - No mixed values have been detected in df_ords.

## 9.3. Addressing missing values in 'df_ords'

In [25]:
# Item 5. Running a check for missing values in df_ords:

df_ords.shape

(3421083, 8)

In [26]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
time_of_order, hour            0
days_since_prior_order    206209
dtype: int64

#### - Findings and explanations for missing values: there are 206,209 missing values for "days_since_prior_order"

In [27]:
# Item 6. Addressing the missing values:

df_ords.describe().apply(lambda s: s.apply('{0:.5f}'.format))

Unnamed: 0       order_id        user_id   order_number  \
count  3421083.00000  3421083.00000  3421083.00000  3421083.00000   
mean   1710541.00000  1710542.00000   102978.20806       17.15486   
std     987581.73983   987581.73982    59533.71779       17.73316   
min          0.00000        1.00000        1.00000        1.00000   
25%     855270.50000   855271.50000    51394.00000        5.00000   
50%    1710541.00000  1710542.00000   102689.00000       11.00000   
75%    2565811.50000  2565812.50000   154385.00000       23.00000   
max    3421082.00000  3421083.00000   206209.00000      100.00000   

      orders_day_of_week time_of_order, hour days_since_prior_order  
count      3421083.00000       3421083.00000          3214874.00000  
mean             2.77622            13.45202               11.11484  
std              2.04683             4.22609                9.20674  
min              0.00000             0.00000                0.00000  
25%              1.00000            10.00000                4.00000  
50%              3.00000            13.00000                7.00000  
75%              5.00000            16.00000               15.00000  
max              6.00000            23.00000               30.00000

In [28]:
    ## Calculating the mean:
    
df_ords.median()

/var/folders/_w/cktj2zv50h34n2spc6h0z5pc0000gp/T/ipykernel_76375/3601841096.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_ords.median()


Unnamed: 0                1710541.0
order_id                  1710542.0
user_id                    102689.0
order_number                   11.0
orders_day_of_week              3.0
time_of_order, hour            13.0
days_since_prior_order          7.0
dtype: float64

In [29]:
    ## Imputing the mean instead of missing values:
    
df_ords['days_since_prior_order'].fillna(7, inplace=True)

In [30]:
    ## Checking if missing values are gone:
    
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
eval_set                  0
order_number              0
orders_day_of_week        0
time_of_order, hour       0
days_since_prior_order    0
dtype: int64

In [31]:
    ## Checking the change in the mean for "days_since_prior_order":
    
df_ords.describe().apply(lambda s: s.apply('{0:.5f}'.format))

Unnamed: 0       order_id        user_id   order_number  \
count  3421083.00000  3421083.00000  3421083.00000  3421083.00000   
mean   1710541.00000  1710542.00000   102978.20806       17.15486   
std     987581.73983   987581.73982    59533.71779       17.73316   
min          0.00000        1.00000        1.00000        1.00000   
25%     855270.50000   855271.50000    51394.00000        5.00000   
50%    1710541.00000  1710542.00000   102689.00000       11.00000   
75%    2565811.50000  2565812.50000   154385.00000       23.00000   
max    3421082.00000  3421083.00000   206209.00000      100.00000   

      orders_day_of_week time_of_order, hour days_since_prior_order  
count      3421083.00000       3421083.00000          3421083.00000  
mean             2.77622            13.45202               10.86681  
std              2.04683             4.22609                8.97852  
min              0.00000             0.00000                0.00000  
25%              1.00000            10.00000                5.00000  
50%              3.00000            13.00000                7.00000  
75%              5.00000            16.00000               15.00000  
max              6.00000            23.00000               30.00000

#### Why I used imputing mean values method: 1. The number of missing values (206,209) was a considerable amount - around 6.02% of the total number of values in the column, so it could not be disregarded. 2. Imputing seemed like a pretty natural thing to do to deal with missing values for number of days between orders, and for missing data comprising 6% of the column data. 3. When choosing between imputing mean or median values, I chose median, since I did not want to raise the mean of the column too much. (It did go down by 0.24803 after imputing - 0.82% of the data range.)

## 9.4. Checking for duplicates in 'df_ords'

In [32]:
# Item 7. Checking for duplicate values in df_ords:

df_ords_dups = df_ords[df_ords.duplicated()]
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, time_of_order, hour, days_since_prior_order]
Index: []

#### - Findings and an explanation for duplicate values: There are no duplicates found in df_ords.

In [33]:
# Item 8. Addressing duplicate values in df_ords: There are no duplicates found in df_ords

#### - Explanation of the menthod: N/A

## 9.5. Exporting 'df_ords' as 'orders_checked'

In [34]:
# Item 9. Exporting df_ords:

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data','orders_checked.csv'))

In [35]:
df_ords.shape

(3421083, 8)